python/python/lance/arrow.py

```diff
@@ -11,14 +11,347 @@
+    def __init__(self, arrow_type, shape):
+        pa.ExtensionType.__init__(
+            self,
+            pa.fixed_shape_tensor(arrow_type, shape),
+            "lance.arrow.fixed_shape_image_tensor",
+        )
+
+    def __arrow_ext_serialize__(self):
+        return b""
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, self, storage_type, serialized):
+        return FixedShapeImageTensorType()
+
+    def __arrow_ext_class__(self):
+        return FixedShapeImageTensorArray
+
+    def __arrow_ext_scalar_class__(self):
+        return FixedShapeImageTensorScalar
+
+
+class ImageURIArray(pa.ExtensionArray):
+    def __repr__(self):
+        return "<lance.arrow.ImageURIArray object at 0x%016x>\n%s" % (
+            id(self),
+            repr(self.to_pylist()),
+        )
+
+    @classmethod
+    def from_uris(cls, uris: Union[pa.StringArray, Iterable[Union[str, Path]]]):
+        """
+        Create an ImageURIArray from a pa.StringArray or an iterable.
+
+        Parameters
+        ----------
+        uris : Union[pa.StringArray, Iterable[Union[str, Path]]]
+
+        Returns
+        -------
+        ImageURIArray
+            Array of image URIs
+
+        Examples
+        --------
+        >>> uris = ["file::///tmp/1.png"]
+        >>> ImageURIArray.from_uris(uris)
+        <lance.arrow.ImageURIArray object at 0x...>
+        ['file::///tmp/1.png']
+        """
+        from urllib.parse import urlparse
+
+        if isinstance(uris, Iterable):
+            uris = pa.array((str(uri) for uri in uris), type=pa.string())
+        else:
+            raise TypeError("Cannot build a ImageURIArray from {}".format(type(uris)))
+
+        for uri in uris:
+            # This is to catch malformed URIs
+            parsed_uri = urlparse(uri.as_py())
+            if not parsed_uri.scheme and not parsed_uri.scheme == "":
+                raise ValueError("URI {} is not a valid URI".format(uri))
+
+        return cls.from_storage(ImageURIType(), uris)
+
+    def read_uris(self):
+        """
+        Read the images from the URIs into memory and return an EncodedImageArray
+
+        Returns
+        -------
+        EncodedImageArray
+            Array of encoded images
+
+        Examples
+        --------
+        >>> import os
+        >>> uris = [os.path.join(os.path.dirname(__file__), "../tests/images/1.png")]
+        >>> uri_array = ImageURIArray.from_uris(uris)
+        >>> uri_array.read_uris()
+        <lance.arrow.EncodedImageArray object at 0x...>
+        [...]
+        """
+        from pyarrow import fs
+
+        images = []
+        for uri in self.storage:
+            filesystem, path = fs.FileSystem.from_uri(uri.as_py())
+            with filesystem.open_input_stream(path) as f:
+                images.append(f.read())
+
+        return EncodedImageArray.from_storage(
+            EncodedImageType(), pa.array(images, type=pa.binary())
+        )
+
+
+class EncodedImageArray(pa.ExtensionArray):
```

Comment:
Worth adding a docstring to the class. We should explain the images may be in any format, and there isn't an internal discriminant marking the format type. The format should be inferred from the image bytes, such as the [PNG signature](https://www.w3.org/TR/2003/REC-PNG-20031110/#5PNG-file-signature) for PNG.

Reply:
Added a docstring.